File: bridge/src/esp32-serial.ts
/**
 * ESP32 Serial Bridge — broadcasts BridgeEvents over USB serial.
 *
 * Detects ESP32 devices (CH340/CP210x) on USB serial ports,
 * opens the port, and sends newline-delimited JSON matching
 * the same protocol as WebSocket.
 *
 * ESP32 side reads lines starting with '{' and passes to Protocol::parseMessage().
 */

import { execSync } from 'child_process';
import { createWriteStream, type WriteStream } from 'fs';
import type { BridgeEvent } from './types.js';
import { debug } from './logger.js';

// Serial port patterns for ESP32 devices
const ESP32_PORT_PATTERNS = [
  /\/dev\/cu\.usbserial-\d+/,   // CH340 (86 Box)
  /\/dev\/cu\.usbmodem\d+/,      // Native USB JTAG (IPS 3.5", Round AMOLED)
  /\/dev\/ttyUSB\d+/,            // Linux CH340
  /\/dev\/ttyACM\d+/,            // Linux native USB
];

// Exclude known non-ESP32 devices
const EXCLUDE_PATTERNS = [
  /Bluetooth/i,
  /WLAN/i,
];

interface SerialConnection {
  port: string;
  stream: WriteStream;
  connected: boolean;
}

let connections: SerialConnection[] = [];
let pollTimer: ReturnType<typeof setInterval> | null = null;

// Events to forward (same subset as Android WS client receives)
const FORWARDED_EVENTS = new Set([
  'state_update',
  'usage_update',
  'sessions_list',
  'connection',
  'timeline_event',
  'timeline_history',
]);

function detectESP32Ports(): string[] {
  try {
    // List serial ports on macOS/Linux
    const platform = process.platform;
    let ports: string[];

    if (platform === 'darwin') {
      const output = execSync('ls /dev/cu.usb* 2>/dev/null || true', {
        encoding: 'utf-8',
        timeout: 3000,
      });
      ports = output.trim().split('\n').filter(Boolean);
    } else if (platform === 'linux') {
      const output = execSync('ls /dev/ttyUSB* /dev/ttyACM* 2>/dev/null || true', {
        encoding: 'utf-8',
        timeout: 3000,
      });
      ports = output.trim().split('\n').filter(Boolean);
    } else {
      return [];
    }

    // Filter to ESP32 patterns, exclude known non-ESP32
    return ports.filter(port => {
      if (EXCLUDE_PATTERNS.some(p => p.test(port))) return false;
      return ESP32_PORT_PATTERNS.some(p => p.test(port));
    });
  } catch {
    return [];
  }
}

function openPort(port: string): SerialConnection | null {
  try {
    // Configure baud rate + disable DTR/RTS to prevent ESP32 reset
    const platform = process.platform;
    if (platform === 'darwin') {
      execSync(`stty -f ${port} 115200 cs8 -cstopb -parenb -hupcl`, { timeout: 3000 });
    } else if (platform === 'linux') {
      execSync(`stty -F ${port} 115200 cs8 -cstopb -parenb -hupcl`, { timeout: 3000 });
    }

    const stream = createWriteStream(port, { flags: 'w' });
    const conn: SerialConnection = { port, stream, connected: true };

    stream.on('error', (err) => {
      debug('ESP32', `Serial error on ${port}: ${err.message}`);
      conn.connected = false;
    });

    stream.on('close', () => {
      debug('ESP32', `Serial port closed: ${port}`);
      conn.connected = false;
    });

    debug('ESP32', `Opened serial port: ${port}`);
    return conn;
  } catch (err: any) {
    debug('ESP32', `Failed to open ${port}: ${err.message}`);
    return null;
  }
}

function sendToConnection(conn: SerialConnection, json: string): void {
  if (!conn.connected) return;
  try {
    conn.stream.write(json + '\n');
  } catch {
    conn.connected = false;
  }
}

/**
 * Start ESP32 serial bridge.
 * Detects USB serial ports and opens connections.
 * Call broadcast() to send events to all connected ESP32 devices.
 */
export function startESP32Serial(): void {
  // Initial detection
  pollForDevices();

  // Poll for new/disconnected devices every 10 seconds
  pollTimer = setInterval(pollForDevices, 10000);

  debug('ESP32', 'Serial bridge started');
}

function pollForDevices(): void {
  const ports = detectESP32Ports();

  // Remove disconnected
  connections = connections.filter(c => {
    if (!c.connected) {
      try { c.stream.end(); } catch { /* ignore */ }
      return false;
    }
    return true;
  });

  // Add new ports
  for (const port of ports) {
    if (!connections.some(c => c.port === port)) {
      const conn = openPort(port);
      if (conn) {
        connections.push(conn);
      }
    }
  }
}

/**
 * Broadcast a BridgeEvent to all connected ESP32 devices via serial.
 */
export function broadcastESP32(event: BridgeEvent): void {
  if (connections.length === 0) return;
  if (!FORWARDED_EVENTS.has(event.type)) return;

  const json = JSON.stringify(event);
  for (const conn of connections) {
    sendToConnection(conn, json);
  }
}

/**
 * Stop ESP32 serial bridge and close all connections.
 */
export function stopESP32Serial(): void {
  if (pollTimer) {
    clearInterval(pollTimer);
    pollTimer = null;
  }
  for (const conn of connections) {
    conn.connected = false;
    try { conn.stream.end(); } catch { /* ignore */ }
  }
  connections = [];
  debug('ESP32', 'Serial bridge stopped');
}

/**
 * Get number of connected ESP32 devices.
 */
export function esp32ConnectionCount(): number {
  return connections.filter(c => c.connected).length;
}

File: apple/AgentDeck/Daemon/Modules/ESP32Serial.swift
#if os(macOS)
// ESP32Serial.swift — USB serial communication with ESP32 devices
// Ported from bridge/src/esp32-serial.ts

import Foundation

/// Manages USB serial connections to ESP32 devices (CH340/CP210x/native USB).
/// Newline-delimited JSON protocol, heartbeat, WiFi provisioning.
actor ESP32Serial {
    // Port detection patterns
    private static let portPatterns: [NSRegularExpression] = {
        ["/dev/cu\\.usbserial-\\d+", "/dev/cu\\.wchusbserial\\d+", "/dev/cu\\.usbmodem\\d+"].compactMap {
            try? NSRegularExpression(pattern: $0)
        }
    }()
    private static let excludePatterns = ["Bluetooth", "WLAN"]

    struct SerialConnection: Identifiable {
        let id = UUID()
        let port: String
        var writeHandle: FileHandle?
        var readHandle: FileHandle?
        var connected = true
        var readBuffer = ""
        var deviceInfo: DeviceInfo?
        var provisionSent = false
    }

    struct DeviceInfo {
        var board: String?
        var version: String?
        var wifiConfigured: Bool?
        var wifiConnected: Bool?
    }

    private var connections: [SerialConnection] = []
    private var pollTask: Task<Void, Never>?
    private var heartbeatTask: Task<Void, Never>?

    nonisolated(unsafe) private var stateProvider: (() -> [String: Any]?)?
    nonisolated(unsafe) private var usageProvider: (() -> [String: Any]?)?
    nonisolated(unsafe) private var initialStateProvider: (() -> [[String: Any]])?
    var onMessage: (@Sendable (String, [String: Any]) -> Void)?

    var connectionCount: Int { connections.filter(\.connected).count }

    nonisolated func setStateProviderFn(_ provider: @escaping () -> [String: Any]?) { stateProvider = provider }
    nonisolated func setUsageProviderFn(_ provider: @escaping () -> [String: Any]?) { usageProvider = provider }
    nonisolated func setInitialStateProviderFn(_ provider: @escaping () -> [[String: Any]]) { initialStateProvider = provider }
    func setOnMessage(_ handler: @escaping @Sendable (String, [String: Any]) -> Void) { onMessage = handler }

    // MARK: - Lifecycle

    func start() {
        pollTask = Task { [weak self] in
            await self?.pollForDevices()
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(10))
                await self?.pollForDevices()
            }
        }

        heartbeatTask = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: .seconds(5))
                await self?.sendHeartbeat()
            }
        }

        DaemonLogger.shared.debug("ESP32", "Serial bridge started")
    }

    func stop() {
        pollTask?.cancel()
        heartbeatTask?.cancel()
        for var conn in connections {
            conn.connected = false
            try? conn.writeHandle?.close()
            try? conn.readHandle?.close()
        }
        connections.removeAll()
        DaemonLogger.shared.debug("ESP32", "Serial bridge stopped")
    }

    // MARK: - Broadcast

    /// Forward events matching SERIAL_FORWARDED_EVENTS to all connected ESP32
    func broadcast(_ event: [String: Any]) {
        guard !connections.isEmpty else { return }
        guard let type = event["type"] as? String,
              Self.serialForwardedEvents.contains(type) else { return }

        let prepared = prepareForSerial(event)
        guard let data = try? JSONSerialization.data(withJSONObject: prepared),
              let json = String(data: data, encoding: .utf8) else { return }

        for i in connections.indices where connections[i].connected {
            sendToConnection(&connections[i], json: json)
        }
    }

    func sendWifiProvisionToAll(_ msg: [String: Any]) -> Int {
        guard let data = try? JSONSerialization.data(withJSONObject: msg),
              let json = String(data: data, encoding: .utf8) else { return 0 }
        var count = 0
        for i in connections.indices {
            guard connections[i].connected, !connections[i].provisionSent else { continue }
            if connections[i].deviceInfo?.wifiConnected == true { continue }
            sendToConnection(&connections[i], json: json)
            connections[i].provisionSent = true
            count += 1
        }
        return count
    }

    // MARK: - Port Detection

    private func detectPorts() -> [String] {
        do {
            let output = try shellSync("ls /dev/cu.usb* 2>/dev/null || true")
            return output.split(separator: "\n").map(String.init).filter { port in
                guard !Self.excludePatterns.contains(where: { port.localizedCaseInsensitiveContains($0) }) else { return false }
                let range = NSRange(port.startIndex..., in: port)
                return Self.portPatterns.contains { $0.firstMatch(in: port, range: range) != nil }
            }
        } catch {
            return []
        }
    }

    private func pollForDevices() {
        // Prune disconnected
        connections.removeAll { !$0.connected }

        let ports = detectPorts()
        for port in ports {
            if !connections.contains(where: { $0.port == port }) {
                if let conn = openPort(port) {
                    connections.append(conn)
                }
            }
        }
    }

    // MARK: - Port Open

    private func openPort(_ port: String) -> SerialConnection? {
        guard let writeHandle = FileHandle(forWritingAtPath: port) else {
            DaemonLogger.shared.debug("ESP32", "Failed to open write: \(port)")
            return nil
        }

        let readHandle = FileHandle(forReadingAtPath: port)

        var conn = SerialConnection(port: port, writeHandle: writeHandle, readHandle: readHandle)

        // Configure baud rate for UART ports (not CDC)
        let isCDC = port.contains("usbmodem")
        if !isCDC {
            _ = try? shellSync("stty -f \(port) 115200 cs8 -cstopb -parenb -hupcl")
        }

        DaemonLogger.shared.debug("ESP32", "Opened: \(port) [\(isCDC ? "CDC" : "UART")]")

        // Request device info
        sendToConnection(&conn, json: #"{"type":"device_info_request"}"#)

        // Send initial state
        if let events = initialStateProvider?() {
            for event in events {
                guard let type = event["type"] as? String,
                      Self.serialForwardedEvents.contains(type) else { continue }
                let prepared = prepareForSerial(event)
                if let data = try? JSONSerialization.data(withJSONObject: prepared),
                   let json = String(data: data, encoding: .utf8) {
                    sendToConnection(&conn, json: json)
                }
            }
        }

        // Start reading in background
        if let readHandle {
            startReading(port: port, handle: readHandle)
        }

        return conn
    }

    private func startReading(port: String, handle: FileHandle) {
        handle.readabilityHandler = { [weak self] fh in
            let data = fh.availableData
            guard !data.isEmpty, let str = String(data: data, encoding: .utf8) else { return }
            Task { await self?.handleReadData(port: port, data: str) }
        }
    }

    private func handleReadData(port: String, data: String) {
        guard let idx = connections.firstIndex(where: { $0.port == port }) else { return }
        connections[idx].readBuffer += data

        while let newlineIdx = connections[idx].readBuffer.firstIndex(of: "\n") {
            let line = String(connections[idx].readBuffer[..<newlineIdx]).trimmingCharacters(in: .whitespaces)
            connections[idx].readBuffer = String(connections[idx].readBuffer[connections[idx].readBuffer.index(after: newlineIdx)...])

            guard line.hasPrefix("{") else { continue }
            guard let jsonData = line.data(using: .utf8),
                  let msg = try? JSONSerialization.jsonObject(with: jsonData) as? [String: Any],
                  let type = msg["type"] as? String else { continue }

            DaemonLogger.shared.debug("ESP32", "← \(port): \(type)")

            if type == "device_info" {
                connections[idx].deviceInfo = DeviceInfo(
                    board: msg["board"] as? String,
                    version: msg["version"] as? String,
                    wifiConfigured: msg["wifiConfigured"] as? Bool,
                    wifiConnected: msg["wifiConnected"] as? Bool
                )
            }

            onMessage?(port, msg)
        }

        // Prevent buffer bloat
        if connections[idx].readBuffer.count > 8192 {
            connections[idx].readBuffer = ""
        }
    }

    // MARK: - Heartbeat

    private func sendHeartbeat() {
        guard !connections.isEmpty else { return }

        if let event = stateProvider?() {
            let prepared = prepareForSerial(event)
            if let data = try? JSONSerialization.data(withJSONObject: prepared),
               let json = String(data: data, encoding: .utf8) {
                for i in connections.indices where connections[i].connected {
                    sendToConnection(&connections[i], json: json)
                }
            }
        }

        if let event = usageProvider?(),
           event["fiveHourPercent"] != nil {
            let prepared = prepareForSerial(event)
            if let data = try? JSONSerialization.data(withJSONObject: prepared),
               let json = String(data: data, encoding: .utf8) {
                for i in connections.indices where connections[i].connected {
                    sendToConnection(&connections[i], json: json)
                }
            }
        }
    }

    // MARK: - Serial Helpers

    private func sendToConnection(_ conn: inout SerialConnection, json: String) {
        guard conn.connected, let handle = conn.writeHandle else { return }
        do {
            try handle.write(contentsOf: Data((json + "\n").utf8))
        } catch {
            conn.connected = false
        }
    }

    /// Strip fields ESP32 doesn't need (reduce payload for small RX buffers)
    private func prepareForSerial(_ event: [String: Any]) -> [String: Any] {
        var e = event
        if event["type"] as? String == "usage_update" {
            e.removeValue(forKey: "ollamaStatus")
            e.removeValue(forKey: "tokenStatus")
            e.removeValue(forKey: "extraUsageEnabled")
            e.removeValue(forKey: "extraUsageMonthlyLimit")
            e.removeValue(forKey: "extraUsageUsedCredits")
            e.removeValue(forKey: "extraUsageUtilization")
            e.removeValue(forKey: "costSpent")
            e.removeValue(forKey: "costLimit")
            e.removeValue(forKey: "sessionPercent")
            e.removeValue(forKey: "resetTime")
            e.removeValue(forKey: "resetDate")
        }
        if event["type"] as? String == "state_update" {
            e.removeValue(forKey: "agentCapabilities")
            e.removeValue(forKey: "billingType")
            e.removeValue(forKey: "remoteUrl")
        }
        return e
    }

    private func shellSync(_ command: String) throws -> String {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/bin/sh")
        process.arguments = ["-c", command]
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = FileHandle.nullDevice
        try process.run()
        process.waitUntilExit()
        return String(data: pipe.fileHandleForReading.readDataToEndOfFile(), encoding: .utf8) ?? ""
    }

    // MARK: - Constants

    static let serialForwardedEvents: Set<String> = [
        "state_update", "usage_update", "sessions_list",
        "connection", "display_state",
        "timeline_event", "timeline_history"
    ]
}
#endif
